### PR TITLE
cli: initial implementation using Click

### PR DESCRIPTION
**Pull request by nharraud**
- Refactors to match the last cookie cutter template, and to make all function names consistent.
- Fixes some bugs.
- Adds templating.jinja_to_mapping function.
- Adds CLI.

Coverage is ok but not perfect (97%).

### CODE REVIEW

File: AUTHORS.rst [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    waive the privileges and immunities granted to it by virtue of its status
+    as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+
+Authors
+=======
+
+es-jsonschema provides a library and a CLI generating Elasticsearch mappings from json-schemas.
+
+- CERN <info@invenio-software.org>
```
Comment (jirikuncar) [inline AUTHORS.rst]: update the author list

File: es_jsonschema/cli.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+
+@click.group()
+def cli():
+    """CLI group."""
+    pass  # pragma: no cover
+
+
+@cli.command('schema_to_mapping')
+@click.argument('schema', type=click.File('rb'))
```
Comment (jirikuncar) [inline es_jsonschema/cli.py]: default stdin

File: es_jsonschema/cli.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+@click.argument('output', type=click.File('wb'))
```
Comment (jirikuncar) [inline es_jsonschema/cli.py]: default stdout

File: es_jsonschema/cli.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    if config:
+        with open(config) as conf:
+            config_instance.load(json.loads(conf.read()))
+
+    mapping = schema_to_mapping(parsed_schema, id, {}, config_instance)
+    output.write(json.dumps(mapping, indent=indent))
+
+
+@cli.command('mapping_to_jinja')
+@click.argument('mapping', type=click.File('rb'))
```
Comment (jirikuncar) [inline es_jsonschema/cli.py]: default stdin

File: es_jsonschema/cli.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+@click.argument('output', type=click.File('wb'))
```
Comment (jirikuncar) [inline es_jsonschema/cli.py]: default stdout

File: es_jsonschema/cli.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+@click.option('--indent', '-i', default=4, type=click.INT,
+              help='Output template indentation step.')
+@click.option('--type', '-t', default='type', type=click.STRING,
+              help='Root type name used in jinja block names.')
+def mapping_to_jinja_cli(mapping, output, indent, type):
+    """Generate jinja template from Elasticsearch mapping."""
+    mapping_str = mapping.read()
+    parsed_mapping = json.loads(mapping_str)
```
Comment (jirikuncar) [inline es_jsonschema/cli.py]: `parsed_mapping = json.load(mapping)`

**Comment by nharraud** [comment]
Good catch
I also replaced other `json.dumps` with `json.dump` and `json.loads` with `json.load` when possible.

**Comment by nharraud** [comment]
@jirikuncar PR ready for review. Only invalid use cases remain untested. I can fix that in another PR.

File: es_jsonschema/mapping.py [hunk 7]
```diff
@@ -264,15 +327,16 @@ def __gen_type_properties(json_schema, path, resolver, config, es_mapping):
             es_mapping['properties'] = es_properties
         # build the elasticsearch mapping corresponding to each json schema
         # property
-        for prop, prop_schema in json_schema['properties'].iteritems():
-            es_properties[prop] = __gen_type_properties(prop_schema,
-                                                        path + '/' + prop,
-                                                        resolver, config,
-                                                        es_properties.get(prop))
+        for prop, prop_schema in iteritems(json_schema['properties']):
+            es_properties[prop] = \
```
Comment (jirikuncar) [inline es_jsonschema/mapping.py]: it's preferable to break line after `(` instead of introducing `\`.

File: es_jsonschema/mapping.py [hunk 8]
```diff
@@ -40,10 +46,48 @@ def __init__(self, *args, **kwargs):
...
+        self.__all_field = True
+        self.__date_detection = True
+        self.__numeric_detection = True
+        self.__date_format = None
+
+    def load(self, config):
+        """Load a configuration dict, overriding current configuration.
+
+        The configuration dict should be of the form:
+        .. highlight:: python
```
Comment (jirikuncar) [inline es_jsonschema/mapping.py]: `.. code-block:: python` ?

File: tests/test_cli.py [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+import jinja2
+import six
+from click.testing import CliRunner
+
+from es_jsonschema.cli import jinja_to_mapping_cli, mapping_to_jinja_cli, \
+    schema_to_mapping_cli
+from es_jsonschema.errors import JsonSchemaSupportError
+
+
+class TestCLI(object):
```
Comment (jirikuncar) [inline tests/test_cli.py]: IMHO you don't need the test class anymore. All methods can be simple functional test cases.

File: es_jsonschema/mapping.py [hunk 10]
```diff
@@ -40,10 +46,48 @@ def __init__(self, *args, **kwargs):
         }
         # json string formats -> elasticsearch type
         self.__formats_map = {}
```
Comment (jirikuncar) [inline es_jsonschema/mapping.py]: be careful with using double underscore for properties (https://docs.python.org/2/tutorial/classes.html#private-variables-and-class-local-references) - they make it hard for subclassing ...

File: es_jsonschema/mapping.py [hunk 11]
```diff
@@ -80,47 +129,60 @@ def get_es_type(self, json_type, json_format=None):
...
     def all_field(self):
-        return getattr(self, '__all_field', True)
+        """Return True if elasticsearch mappings should have an "all" field."""
+        return self.__all_field
 
     @all_field.setter
     def all_field(self, enabled):
+        """Enable/Disable "all" field generation in elasticsearch mappings."""
         self.__all_field = enabled
         return self
```
Comment (jirikuncar) [inline es_jsonschema/mapping.py]: no return on setter

File: es_jsonschema/mapping.py [hunk 12]
```diff
@@ -80,47 +129,60 @@ def get_es_type(self, json_type, json_format=None):
 
     @property
     def date_format(self):
-        return getattr(self, '__date_format', None)
+        """Return the date format used in elasticsearch mappings."""
+        return self.__date_format
```
Comment (jirikuncar) [inline es_jsonschema/mapping.py]: Avoid using Java style code with setters and getters unless you are doing something special in your getter/setter. Enjoy Python world!

**Comment by nharraud** [comment]
@jirikuncar Fixed and added an example

File: tests/test_cli.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+    """
+    if result.exception:
+        traceback.print_exception(*result.exc_info)
+    assert not result.exception
+
+
+def test_schema_to_mapping():
+    """Test schema_to_mapping."""
+    json_schema = {
+        'id': 'http://some.site/root_schema#',
```
Comment (jirikuncar) [inline tests/test_cli.py]: `https://example.org/`

File: tests/test_generate_mapping.py [hunk 14]
```diff
@@ -3,110 +3,320 @@
...
+    schema_to_mapping
 
-from es_jsonschema.mapping import \
-    ElasticMappingGeneratorConfig, generate_type_mapping
 
-import pytest
+def test_simple_properties():
+    """Test generation of a very simple mapping"""
+    json_schema = {
+        'id': 'http://some.site/root_schema#',
```
Comment (jirikuncar) [inline tests/test_generate_mapping.py]: `https://example.org/`

**Comment by nharraud** [comment]
@jirikuncar squashed in one commit as requested.

**Comment by jirikuncar** [comment]
@nharraud please run `kwalitee check message`

**Comment by nharraud** [comment]
@jirikuncar fixed and travis finished at last.

**Pull request merged**